### PR TITLE
{AKS} Graph module migration

### PR DESCRIPTION
**Pull request by FumingZhang**
**Related command**

`az aks create`

**Description**

- Graph module migration.
  - Remove most dependencies on graph SDK.
  - Only keep the functions that query object id by app id, which is used by role assignments.
  - Keep the graph SDK related calls used by the `acs` commands unchanged, and plan to remove the deprecated `acs` command group as a whole during the next breaking change window to eliminate the dependency on the legacy graph SDK.
  - Split the global import of legacy graph SDK in `custom.py` into each function, so that even if CLI removes the dependency, import errors should not occur without using these deprecated commands.
- After the changes in this PR, a service principal will no longer be implicitly created for the user when the user has not provided both app id and secret (through options `--service-principal` and `--client-secret`), instead a `RequiredArgumentMissingError` exception would be raised to prompt the users to create it by themselves.
  - Although the default behavior of aks create is to use managed identity as the cluster's identity, users can still set service principal as the cluster's identity by specifying the --service-principal and --client-secret options. And there is still a [piece of code](https://github.com/Azure/azure-cli/blob/62a01da2b5f2de505e37ed8f444e06c5b8dd5f09/src/azure-cli/azure/cli/command_modules/acs/managed_cluster_decorator.py#L1123) that helps create an [application and service principal](https://github.com/Azure/azure-cli/blob/62a01da2b5f2de505e37ed8f444e06c5b8dd5f09/src/azure-cli/azure/cli/command_modules/acs/_graph.py#L188) when the user only provides --client-secret. But with this graph module migration, the previously used Azure AD Graph API will be replaced by the Microsoft Graph API (see [doc](https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-overview), which does not support specifying password credential when creating [application](https://docs.microsoft.com/en-us/graph/api/application-post-applications?view=graph-rest-beta&tabs=http) and [service principals](https://docs.microsoft.com/en-us/graph/api/serviceprincipal-post-serviceprincipals?view=graph-rest-beta&tabs=http), which makes it impossible for us to maintain the previous behavior.
  - Therefore, I propose that when the user does not provide the complete service principal information (i.e., provide the app id and secret at the same time), instead of trying to create an application and service principal for the users, an error is thrown to prompt the users to create it by themselves. In my opinion, this change will not be noticeable to most users of aks, because the default create behavior is to use managed identity, and users who use service principal as identity will generally provide both id and secret. In addition, after this change, we will basically remove the legacy dependency on the graph module on the CLI side, making the code more streamlined and conducive to subsequent maintenance.

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

**Comment by FumingZhang** [comment]
@weinong, @jiasli Would appreciate if you could help review. Thanks in advance!

File: src/azure-cli/azure/cli/command_modules/acs/_graph.py [hunk 1]
```diff
@@ -3,187 +3,39 @@
...
+        result = list(client.user_list(filter="userPrincipalName eq '{}'".format(assignee)))
+    # find by spn
     if not result:
-        result = list(client.service_principals.list(filter="servicePrincipalNames/any(c:c eq '{}')".format(assignee)))
-    if not result:  # assume an object id, let us verify it
-        result = _get_object_stubs(client, [assignee])
+        result = list(client.service_principal_list(filter="servicePrincipalNames/any(c:c eq '{}')".format(assignee)))
+    # find by display name
+    if not result:
+        result = list(client.service_principal_list(filter="displayName eq '{}'".format(assignee)))
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/_graph.py:26]: this is a new logic? (by displayName?)

File: src/azure-cli/azure/cli/command_modules/acs/_graph.py [hunk 2]
```diff
@@ -3,187 +3,39 @@
...
-    logger.info("Finished service principal creation")
-    return service_principal, aad_session_key
-
-
-def _create_client_secret():
-    # Add a special character to satisfy AAD SP secret requirements
-    special_char = "$"
-    client_secret = binascii.b2a_hex(os.urandom(10)).decode("utf-8") + special_char
-    return client_secret
+    return result[0]["id"]
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/_graph.py:35]: will it possible that the result is not none, but length is zero.

Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/_graph.py:35]: No, if this is an empty list, the if statement (empty list is converted to False) at the two lines above will raise an exception.

File: src/azure-cli/azure/cli/command_modules/acs/_graph.py (hunk 1 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/_graph.py:26]: Yes, I added it. Now updated the comments with examples.

**Review by jiasli — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Comment by jiasli** [comment]
Linking https://github.com/Azure/azure-cli/issues/22174

**Pull request merged**